### ISSUE TITLE
[Epic] UAF / Early-Deallocation: CQ & DMA Buffer Lifetime Safety Audit

### ISSUE THREAD

**Issue by Riddy21** [tenstorrent/tt-metal#45230]
## Overview

A class of **use-after-free (UAF) / early-deallocation bugs** exists across the command queue and DMA memory management paths in tt-metal. The common root cause is: raw `void*` pointers (or stack-local references) are stored in async descriptor/command structures and consumed by background threads — with no shared ownership or lifetime guarantee on the pointed-to buffer. When the caller frees the buffer before the async work completes, the background thread reads/writes freed memory.

This was first reported and partially fixed in PR #44207. This epic tracks all similar patterns found in the subsequent audit.

---

## Root Cause Pattern

```
Caller provides raw void* dst
         ↓
Stored in descriptor struct (ReadCoreDataDescriptor, ReadBufferDescriptor, etc.)
         ↓
Descriptor enqueued into async queue (completion_queue_reads_, read_descriptors_)
         ↓
Caller returns (blocking=false) — buffer may be freed
         ↓
Background thread pops descriptor, dereferences void* dst → USE-AFTER-FREE
```

---

### Confirmed Bugs — Raw Pointer in Async Descriptor

| # | Issue | File | Pattern | Status |
|---|---|---|---|---|
| 1 | #43638 | `tt_metal/impl/buffers/dispatch.cpp` | UAF in `Tensor::cpu(blocking=False)` | Fixed by PR #44207 |
| 2 | #45227 | `tt_metal/impl/device/dispatch.hpp:16` | `ReadCoreDataDescriptor::dst` raw void* in async completion queue | Open |
| 3 | #45228 | `tt_metal/impl/buffers/dispatch.hpp:42` | `ReadBufferDescriptor::dst` raw void* in single-device CQ reader | Open |
| 4 | #45229 | `tt_metal/impl/profiler/profiler.cpp:1223,1270,1343` | Profiler calls `enqueue_read_shard_from_core(blocking=false)` | Open |

### Confirmed Bugs — Detach + TT_THROW UAF on Stack Locals

| # | Issue | File | Pattern | Status |
|---|---|---|---|---|
| 5 | #44451 | `tt_metal/distributed/mesh_socket_utils.hpp:89-117` | `execute_with_timeout`: thread.detach() + TT_THROW destroys captured stack locals | Open |
| 6 | #44452 | `tt_metal/fabric/topology_mapper.cpp:80-117` | Same `execute_with_timeout` pattern | Open |

### Flaky Tests / Possible Symptoms

| # | Issue | Notes | Status |
|---|---|---|---|
| 7 | #43725 | `TestAsyncRuntimeAllocatedBuffers` data mismatch — likely multi-CQ event sync UAF | Open |
| 8 | #39650 | Double free in slow dispatch eth tests (BH) | Open |

### Open PR Fixing Part of This

| # | PR | Description |
|---|---|---|
| — | #44207 | Adds `MemoryPin` to `MeshBufferReadDescriptor` for mesh D2H reads (partial fix) |

---

### For raw pointer in async descriptor (#45227, #45228, #45229):
Replace `void* dst` in `ReadCoreDataDescriptor` and `ReadBufferDescriptor` with a `MemoryPin` (RAII ref-counting wrapper), mirroring the approach in PR #44207 for `MeshBufferReadDescriptor`. The pin keeps the host buffer alive until the async memcpy thread completes.

### For detach+throw UAF (#44451, #44452):
Replace `thread.detach()` + `TT_THROW` with a cancellation-flag pattern:
```cpp
std::atomic<bool> timed_out{false};
auto t = std::thread([&]{ ...; timed_out.store(true); });
t.join();  // always join; use timed_out to detect timeout
if (timed_out) TT_THROW("Timed out...");
```

---

## Files to Audit

The following files contain the affected structs and their async consumers:

```
tt_metal/impl/device/dispatch.hpp          — ReadCoreDataDescriptor
tt_metal/impl/buffers/dispatch.hpp         — ReadBufferDescriptor, BufferReadDispatchParams
tt_metal/distributed/fd_mesh_command_queue.cpp — enqueue/dequeue sites, completion queue reader thread
tt_metal/distributed/mesh_command_queue_base.cpp — enqueue_write_shard_to_sub_grid
tt_metal/distributed/mesh_socket_utils.hpp — execute_with_timeout
tt_metal/fabric/topology_mapper.cpp        — execute_with_timeout
tt_metal/impl/profiler/profiler.cpp        — non-blocking core reads
```

---

cc @Riddy21
